### PR TITLE
Deprecate old streams

### PR DESCRIPTION
**Pull request by parrt**
WIP update doc, deprecate old interface

### CODE REVIEW

File: doc/unicode.md [hunk 1]
```diff
@@ -1,12 +1,29 @@
...
-(code points up to `U+FFFF`).
+Prior to ANTLR 4.7, generated lexers only supported part of the Unicode standard (code points up to `U+FFFF`). As of ANTLR 4.7, the lexers in all language runtimes support the full range of Unicode code points up to `U+10FFFF`, as
+long as the input `CharStream` is opened using `CharStreams.fromPath()`, `CharStreams.fromFileName()`, etc...
+or the equivalent method for your runtime's language. 
 
-With ANTLR 4.7 and later, lexers as well as all languages' runtimes
-support the full range of Unicode code points up to `U+10FFFF`, as
-long as the input `CharStream` is opened using `CharStreams.fromPath()`
-or the equivalent method for your runtime's language.
+The deprecated `ANTLRInputStream` and `ANTLRFileStream` APIs only support Unicode code points up to `U+FFFF`.
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Mention this is for Java.

File: doc/unicode.md [hunk 2]
```diff
@@ -1,12 +1,29 @@
...
+
+A big shout out to Ben Hamilton (github bhamiltoncx) for his superhuman
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Aww shucks

File: doc/unicode.md [hunk 3]
```diff
@@ -1,12 +1,29 @@
...
+efforts across all targets to get true Unicode 3.1 support for U+10FFFF.
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Actually, it was Unicode 2.0 way back in 1996! http://unicode.org/faq/utf_bom.html#gen0

File: doc/unicode.md [hunk 4]
```diff
@@ -37,35 +54,83 @@ NOT_CYRILLIC : [\P{Script=Cyrillic}] ;
...
-`ANTLRFileStream` APIs only support Unicode code points up to `U+FFFF`.
+Code for **4.6** looked like this:
 
-The existing `TestRig` command-line interface supports all Unicode
-code points.
 
-## Other Targets
-Other language targets usually have their `ANTLRInputStream` extended to support the full Unicode range. See the target documentation for supported input encodings (e.g. UTF-8) and other related details.
+```java
+CharStream input = new ANTLRFileStream("myinputfile")
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Add missing `;`

File: doc/unicode.md [hunk 5]
```diff
@@ -37,35 +54,83 @@ NOT_CYRILLIC : [\P{Script=Cyrillic}] ;
...
-  CommonTokenStream tokens = new CommonTokenStream(lexer);
-  tokens.fill();
-  for (Token token : tokens.getTokens()) {
-    System.out.println("Got token: " + token.toString());
-  }
-}
+Or, if you'd like to specify the file encoding:
+
+```java
+CharStream input = CharStreams.fromFileName("inputfile", StandardCharsets.UTF_16);
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Let's use something more realistic, like `Charset.forName("windows-1252")` (the default Windows text encoding for the US and Western Europe).

File: doc/unicode.md [hunk 6]
```diff
@@ -37,35 +54,83 @@ NOT_CYRILLIC : [\P{Script=Cyrillic}] ;
...
+the characters out of the buffer into the token. That defeats the purpose
+of unbuffered input. See the [ANTLR 4 book](https://www.amazon.com/Definitive-ANTLR-4-Reference/dp/1934356999) "13.8 Unbuffered Character and Token Streams". Unbuffered streams are primarily
+useful for processing infinite streams *during the parse* and require that you manually buffer characters. Use `UnbufferedCharStream` and `UnbufferedTokenStream`.
+
+```java
+CharStream input = new UnbufferedCharStream(is);
CSVLexer lex = new CSVLexer(input);
// copy text out of sliding buffer and store in tokens lex.setTokenFactory(new CommonTokenFactory(true));
TokenStream tokens = new UnbufferedTokenStream<CommonToken>(lex); CSVParser parser = new CSVParser(tokens); parser.setBuildParseTree(false);
parser.file();
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Fix the ^Ms maybe?

File: doc/unicode.md [hunk 7]
```diff
@@ -37,35 +54,83 @@ NOT_CYRILLIC : [\P{Script=Cyrillic}] ;
...
+constructing tokens to explicitly set the text. This is useful for cases
+where the input stream might not be able to provide arbitrary substrings
+of text from the input after the lexer creates a token (e.g. the
+implementation of `CharStream.getText` in
+`UnbufferedCharStream` throws an
+`UnsupportedOperationException`). Explicitly setting the token text
+allows `Token.getText` to be called at any time regardless of the
+input stream implementation.
+
+*Currently, only Java and C# have these unbuffered streams implemented*.
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Oops, didn't mean to mislead — Java, C#, and C++ all have unbuffered streams today. (C++ just didn't need any changes to support Unicode code points.)

File: runtime/Java/src/org/antlr/v4/runtime/UnbufferedCharStream.java [hunk 8]
```diff
@@ -153,25 +156,31 @@ protected int fill(int n) {
 				int c = nextChar();
 				if (c > Character.MAX_VALUE || c == IntStream.EOF) {
 					add(c);
-				} else {
+				}
+				else {
```
Comment (bhamiltoncx) [inline runtime/Java/src/org/antlr/v4/runtime/UnbufferedCharStream.java]: So sorry!

File: doc/unicode.md (hunk 3 above)
Comment (parrt) [inline doc/unicode.md]: wow!

File: doc/unicode.md (hunk 5 above)
Comment (parrt) [inline doc/unicode.md]: ah. good one!

File: doc/unicode.md (hunk 7 above)
Comment (parrt) [inline doc/unicode.md]: Fixed :)

File: runtime/Java/src/org/antlr/v4/runtime/UnbufferedCharStream.java (hunk 8 above)
Comment (parrt) [inline runtime/Java/src/org/antlr/v4/runtime/UnbufferedCharStream.java]: :)

File: doc/unicode.md [hunk 9]
```diff
@@ -37,35 +54,89 @@ NOT_CYRILLIC : [\P{Script=Cyrillic}] ;
...
-`ANTLRFileStream` APIs only support Unicode code points up to `U+FFFF`.
+Code for **4.6** looked like this:
 
-The existing `TestRig` command-line interface supports all Unicode
-code points.
 
-## Other Targets
-Other language targets usually have their `ANTLRInputStream` extended to support the full Unicode range. See the target documentation for supported input encodings (e.g. UTF-8) and other related details.
+```java
+CharStream input = new ANTLRFileStream("myinputfile");
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Mention this didn't actually use UTF-8 by default, but actually depended on the calling environment's default (`$LANG` etc.)

File: doc/unicode.md [hunk 10]
```diff
@@ -37,35 +54,89 @@ NOT_CYRILLIC : [\P{Script=Cyrillic}] ;
...
+JavaLexer lexer = new JavaLexer(input);
+CommonTokenStream tokens = new CommonTokenStream(lexer);
+```
 
-# Java Example
+Code for **4.7** looks like this:
 
-If you have generated a lexer named `UnicodeLexer`:
+```java
+CharStream input = CharStreams.fromFileName("inputfile");
```
Comment (bhamiltoncx) [inline doc/unicode.md]: Mention that the default is now UTF-8.

**Pull request merged**